File: ecobertura.core/src/ecobertura/core/launching/LaunchDelegateRetriever.java
package ecobertura.core.launching;

import java.util.Arrays;
import java.util.HashSet;
import java.util.logging.Logger;


import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IConfigurationElement;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.core.runtime.Status;
import org.eclipse.debug.core.DebugPlugin;
import org.eclipse.debug.core.ILaunch;
import org.eclipse.debug.core.ILaunchConfiguration;
import org.eclipse.debug.core.ILaunchConfigurationType;
import org.eclipse.debug.core.ILaunchDelegate;
import org.eclipse.debug.core.ILaunchManager;
import org.eclipse.debug.core.model.ILaunchConfigurationDelegate;
import org.eclipse.debug.core.model.ILaunchConfigurationDelegate2;

import ecobertura.core.CorePlugin;

class LaunchDelegateRetriever {
	private static final Logger logger = Logger.getLogger("ecobertura.core.launching"); //$NON-NLS-1$
	
	private IConfigurationElement config;
	
	static LaunchDelegateRetriever fromExtensionConfig(IConfigurationElement config) {
		return new LaunchDelegateRetriever(config);
	}
	
	private LaunchDelegateRetriever(IConfigurationElement config) {
		this.config = config;
	}
	
	ILaunchConfigurationDelegate2 getDelegate() throws CoreException {
		String launchTypeName = config.getAttribute("type"); //$NON-NLS-1$
		ILaunchConfigurationType type = getLaunchConfigurationType(launchTypeName);
		return getLaunchDelegateFor(type);
	}

	private ILaunchConfigurationType getLaunchConfigurationType(String launchTypeName) throws CoreException {
		ILaunchConfigurationType type = 
			DebugPlugin.getDefault().getLaunchManager().getLaunchConfigurationType(launchTypeName);
		if (type == null) {
			throw new CoreException(new Status(Status.ERROR, CorePlugin.PLUGIN_ID, 
					String.format("unknown launch configuration type %s", launchTypeName))); //$NON-NLS-1$
		}
		return type;
	}

	private ILaunchConfigurationDelegate2 getLaunchDelegateFor(ILaunchConfigurationType type) throws CoreException {
		ILaunchDelegate[] delegatesForType = type.getDelegates(new HashSet<String>(
				Arrays.asList(ILaunchManager.RUN_MODE)));
		if (delegatesForType.length == 0) {
			throw new CoreException(new Status(Status.ERROR, CorePlugin.PLUGIN_ID, 
					String.format("no delegate for %s found", type.getName()))); //$NON-NLS-1$
		}
		if (delegatesForType.length > 1) {
			logger.warning(String.format("more than one launch delegate found for %s",  //$NON-NLS-1$
					type.getName()));
		}
		return convertToType2Delegate(delegatesForType[0].getDelegate());
	}

	private ILaunchConfigurationDelegate2 convertToType2Delegate(
			ILaunchConfigurationDelegate delegate) {
		
		if (delegate instanceof ILaunchConfigurationDelegate2) {
			return (ILaunchConfigurationDelegate2) delegate;
		}
		return adaptToType2Delegate(delegate);
	}

	private ILaunchConfigurationDelegate2 adaptToType2Delegate(
			final ILaunchConfigurationDelegate delegate) {
		
		return new ILaunchConfigurationDelegate2() {
			@Override
			public boolean buildForLaunch(ILaunchConfiguration configuration,
					String mode, IProgressMonitor monitor) throws CoreException {
				return true;
			}

			@Override
			public boolean finalLaunchCheck(ILaunchConfiguration configuration,
					String mode, IProgressMonitor monitor) throws CoreException {
				return true;
			}

			@Override
			public ILaunch getLaunch(ILaunchConfiguration configuration,
					String mode) throws CoreException {
				return null;
			}

			@Override
			public boolean preLaunchCheck(ILaunchConfiguration configuration,
					String mode, IProgressMonitor monitor) throws CoreException {
				return true;
			}

			@Override
			public void launch(ILaunchConfiguration configuration, String mode,
					ILaunch launch, IProgressMonitor monitor)
					throws CoreException {
				delegate.launch(configuration, mode, launch, monitor);
			}
		};
	}


}

File: ecobertura.core/src/main/scala/ecobertura/core/launching/LaunchDelegateRetriever.scala
/*
 * This file is part of eCobertura.
 * 
 * Copyright (c) 2009, 2010 Joachim Hofer
 * All rights reserved.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */
package ecobertura.core.launching

import java.util.logging.Logger

import scala.collection.JavaConversions._
import scala.collection.mutable.HashSet

import org.eclipse.core.runtime._
import org.eclipse.debug.core._
import org.eclipse.debug.core.model.ILaunchConfigurationDelegate
import org.eclipse.debug.core.model.ILaunchConfigurationDelegate2

import ecobertura.core.CorePlugin

object LaunchDelegateRetriever {
  def fromExtensionConfig(config: IConfigurationElement) = new LaunchDelegateRetriever(config)
}

class LaunchDelegateRetriever private (config: IConfigurationElement) {
  val logger = Logger.getLogger("ecobertura.core.launching") //$NON-NLS-1$

  def getDelegate = {
    val launchTypeName = config.getAttribute("type") //$NON-NLS-1$
    val launchType = getLaunchConfigurationType(launchTypeName)

    getLaunchDelegateFor(launchType)
  }

  private def getLaunchConfigurationType(launchTypeName: String) = {
    val launchType = DebugPlugin.getDefault.getLaunchManager.getLaunchConfigurationType(
        launchTypeName)
    if (launchType == null) {
      throw new CoreException(new Status(IStatus.ERROR, CorePlugin.pluginId, 
          "unknown launch configuration type %s".format(launchTypeName))) //$NON-NLS-1$
    }
    launchType
  }

  private def getLaunchDelegateFor(launchType: ILaunchConfigurationType) = {
    val delegatesForType = launchType.getDelegates(HashSet(ILaunchManager.RUN_MODE))
    if (delegatesForType.isEmpty) {
      throw new CoreException(new Status(IStatus.ERROR, CorePlugin.pluginId,
          "no delegate for %s found".format(launchType.getName)))
    }
    convertToType2Delegate(delegatesForType.head.getDelegate)
  }

  private def convertToType2Delegate(delegate: ILaunchConfigurationDelegate) = {
    if (delegate.isInstanceOf[ILaunchConfigurationDelegate2]) 
      delegate.asInstanceOf[ILaunchConfigurationDelegate2]
    else 
      adaptToType2Delegate(delegate)
  }

  private def adaptToType2Delegate(delegate: ILaunchConfigurationDelegate) = {
    new ILaunchConfigurationDelegate2() {
      override def buildForLaunch(configuration: ILaunchConfiguration, mode: String,
          monitor: IProgressMonitor) = true

      override def finalLaunchCheck(configuration: ILaunchConfiguration, mode: String,
          monitor: IProgressMonitor) = true

      override def getLaunch(configuration: ILaunchConfiguration, mode: String) = null

      override def preLaunchCheck(configuration: ILaunchConfiguration, mode: String,
          monitor: IProgressMonitor) = true

      override def launch(configuration: ILaunchConfiguration, mode: String, launch: ILaunch,
          monitor: IProgressMonitor) = delegate.launch(configuration, mode, launch, monitor)
    }
  }
}